app/assets/javascripts/angular/controllers/mobile/messages.js

```diff
@@ -0,0 +1,197 @@
+        // Add the new message to the existing messages, update the latest.
+        conversationToUpdate[0].latest_message = savedMessage;
+        conversationToUpdate[0].messages.push(savedMessage);
+      } else {
+        // If this is a new conversation, add the new partner.
+        var conversationPartner = $.grep($scope.teams, function(team) {
+          return team.id == conversationPartnerId;
+        });
+        $scope.messages.push({
+          conversation_partner: conversationPartner[0],
+          latest_message: savedMessage,
+          messages: [savedMessage]
+        });
+      }
+
+      // Sort now that the dates have changed.
+      $scope.messages = _.sortBy($scope.messages, function(conversation) {
+        return conversation.latest_message.updated_at;
+      }).reverse();
+    }
+
+    // Message Filtering
+    $scope.filterFn = function(conversation) {
+      var filter = $scope.messageFilter.id;
+
+      if (filter === 0) {
+        return true;
+      } else if (filter === 1) {
+        return conversation.latest_message.sender_id !== $scope.myCountryId;
+      } else if (filter === 2) {
+        return conversation.latest_message.sender_id === $scope.myCountryId;
+      } else {
+        return true;
+      }
+    };
+
+    $scope.initializeMessages();
+
+    // Group individual messages into Conversations.
+    var groupMessages = function(messages) {
+      var myCountryId = $scope.myCountryId;
+      var teams = $scope.teams;
+
+      return _.chain(messages)
+      .groupBy(function(message) {
+        // Group messages by who the conversation is actually with.
+        return message.sender_id === myCountryId ? message.recipient_id : message.sender_id;
+      })
+      .map(function(messages) {
+        // Get the Team name that the conversation is with.
+        var conversation_partner_id = messages[0].sender_id === myCountryId ? messages[0].recipient_id : messages[0].sender_id;
+        var conversation_partner = $.grep(teams, function(team){ return team.id == conversation_partner_id; });
+
+        // Get the latest message.
+        var updated = _.chain(messages)
+          .sortBy(function(message) { return message.updated_at; })
+          .last()
+          .value()
+
+        // Turn this array of messages into a more complex object containing all of these things.
+        return {
+          conversation_partner: conversation_partner[0],
+          latest_message: updated,
+          messages: messages
+        };
+      })
+      .sortBy(function(messages) {
+        return messages.latest_message.updated_at;
+      })
+      .reverse()
+      .value()
+    }
+
+    var apiCall = function() {
+      $http.get('/api/messages_data').then(
+      function successCallback(response) {
+        $('#connection-error').hide();
+        var messages = response['data']['result']['messages'];
+        var formattedMessages = groupMessages(messages);
+
+        if ($scope.messages === formattedMessages) {
```

Comment:
This is the part that I still need to adjust. Right now, even if there are no new messages, it'll still overwrite everything, which is not great.